soroban-env-common/src/env_val.rs

```diff
@@ -12,22 +13,27 @@ use super::{
     Env,
 };
 
-pub trait IntoVal<E: Env, V>: Sized {
-    fn into_val(self, env: &E) -> V;
```

Comment:
I'm suspicious of getting rid of `IntoVal`. The SDK uses `IntoVal` in a few locations for converting into types that should always succeed, such as converting from a fixed length array into a `RawVal`. Maps and Vecs require their T, K, and V types to be `IntoVal<E, RawVal>`.

Will we use `TryIntoVal` instead now? What will we do with the failure case?

Will we create a new trait for that into rawval case instead?

Reply:
So .. the "always succeeds" part here is .. the problematic part. All these impls are currently parameterized over `E:Env` which allows you to call host methods that "always succeed", except in reality they can fail (at minimum "out of gas" but also other things like bad handles -- stuff that turns into a `HostError` and is then escalated to a panic by `impl Env for CheckedEnv` or a VM trap by the dispatch function). When we get rid of the `impl Env for CheckedEnv` (or at least remove it form visibility in `Host`) we will either have to duplicate all of them, but in fallible form, for the `CheckedEnv` case (surfacing errors) or else certain conversions (eg. the current u64-to-object injection) will just disappear from the `Host` context (eg. won't be usable in the builtin contract).

What I propose happen (and am doing groundwork for here) is that _every_ conversion that uses the environment is _always_ considered fallible in the signatures-and-traits sense (returning a `Result<>`), and we merge `Env` and `CheckedEnv`, making the `Env` trait one where all host functions return `Result<>` with an associated type `Env::Error`, and `Guest::Error = Infallible`. That'll be a subsequent PR though.